brave/src/test/java/brave/features/handler/RedactingFinishedSpanHandlerTest.java

```diff
@@ -59,40 +63,69 @@ static String maybeUpdateValue(String value) {
     }
   }
 
-  List<Span> spans = new ArrayList<>();
+  BlockingQueue<Span> spans = new LinkedBlockingQueue<>();
+  FinishedSpanHandler redacter = new FinishedSpanHandler() {
+    @Override public boolean handle(TraceContext context, MutableSpan span) {
+      span.forEachTag(ValueRedactor.INSTANCE);
+      span.forEachAnnotation(ValueRedactor.INSTANCE);
+      return true;
+    }
+
+    @Override public boolean supportsOrphans() {
+      return true; // also redact data leaked by bugs
+    }
+  };
+
   Tracing tracing = Tracing.newBuilder()
-    .addFinishedSpanHandler(new FinishedSpanHandler() {
-      @Override public boolean handle(TraceContext context, MutableSpan span) {
-        span.forEachTag(ValueRedactor.INSTANCE);
-        span.forEachAnnotation(ValueRedactor.INSTANCE);
-        return true;
-      }
-    })
+    .addFinishedSpanHandler(redacter)
     .spanReporter(spans::add)
     .build();
 
   @After public void close() {
     tracing.close();
   }
 
-  @Test public void showRedaction() {
+  @Test public void showRedaction() throws Exception {
     ScopedSpan span = tracing.tracer().startScopedSpan("auditor");
     try {
       span.tag("a", "1");
-      span.tag("b", "912-23-1433");
-      span.annotate("SSN=912-23-1433");
+      span.tag("b", "4121-2319-1483-3421");
+      span.annotate("cc=4121-2319-1483-3421");
       span.tag("c", "3");
     } finally {
       span.finish();
     }
 
-    assertThat(spans.get(0).tags()).containsExactly(
+    Span finished = spans.take();
+    assertThat(finished.tags()).containsExactly(
       entry("a", "1"),
-      // SSN tag was nuked
+      // credit card tag was nuked
       entry("c", "3")
     );
-    assertThat(spans.get(0).annotations()).flatExtracting(Annotation::value).containsExactly(
-      "SSN=xxx-xx-xxxx"
+    assertThat(finished.annotations()).flatExtracting(Annotation::value).containsExactly(
+      "cc=xxx-xx-xxxx"
+    );
+
+    // Leak some data by adding a tag using the same context after the span was finished.
+    tracing.tracer().toSpan(span.context()).tag("d", "cc=4121-2319-1483-3421");
+    span = null; // Orphans are via GC, to test this, we have to drop any reference to the context
+    blockOnGC();
+
+    // GC only clears the reference to the leaked data. Normal tracer use implicitly handles orphans
+    tracing.tracer().nextSpan().start().finish();
+
+    Span leaked = spans.take();
+    assertThat(leaked.tags()).containsExactly(
+      // credit card tag was nuked
+      entry("d", "cc=xxx-xx-xxxx")
     );
+    assertThat(leaked.annotations()).flatExtracting(Annotation::value).containsExactly(
+      "brave.flush"
+    );
+  }
+
+  static void blockOnGC() throws InterruptedException {
+    System.gc();
+    Thread.sleep(200L);
```

Comment:
To make sure the test doesn't take longer than it needs to, can consider `await` on a weak reference becoming null

Reply:
I'm moving this and similar to a new help-wanted issue as this isn't new, but would be nice.

Reply:
https://github.com/openzipkin/brave/issues/975